tls/s2n_send.c

```diff
@@ -167,10 +168,19 @@ ssize_t s2n_sendv_with_offset(struct s2n_connection *conn, const struct iovec *b
                 cbcHackUsed = 1;
             }
         }
-
-        /* Write and encrypt the record */
+        
         GUARD(s2n_stuffer_rewrite(&conn->out));
-        GUARD(s2n_record_writev(conn, TLS_APPLICATION_DATA, bufs, count,
+        if ((conn->secure.cipher_suite == &s2n_tls13_aes_128_gcm_sha256) |
+                (conn->secure.cipher_suite == &s2n_tls13_aes_256_gcm_sha384)) {
+            GUARD(s2n_check_key_limits(conn, to_write));
+            if (conn->key_update_pending) {
+                GUARD(s2n_post_handshake_send(conn, TLS_KEY_UPDATE));
```

Comment:
What does this s2n_post_handshake_send method do? It can currently only send TLS_KEY_UPDATE, because we call it inside the "is this aes-gcm" and "conn->key_update_pending==true" checks.

You could just delete s2n_post_handshake_send and call s2n_key_update_send directly, but I would suggest moving the checks inside s2n_key_update_send. So we always call s2n_post_handshake_send, which always calls send for every post_handshake message, and then the implementation of the message itself decides whether anything needs to be written. That keeps the logic out of sendv. You could probably also make s2n_post_handshake_send handle the flush + rewrite if anything is written.

Reply:
Yeah, I agree this is awkward. I chose to do it this way because I didn't like the idea of calling a send function where there is a possibility that we do not send anything because the limits haven't been reached. But if you think that is more readable, I'm fine with switching it.